src/graphql/graphql-api.php

```diff
@@ -0,0 +1,245 @@
+					array_values( $sourced_block['attributes'] )
+				);
+			}
+		}
+
+		return $sourced_block;
+	}
+
+	/**
+	 * Flatten the inner blocks, no matter how many levels of nesting is there.
+	 *
+	 * @param array $inner_blocks the inner blocks in the block.
+	 * @param array $filter_options Options to filter using, if any.
+	 * @param array $flattened_blocks the flattened blocks that's built up as we go through the inner blocks.
+	 * 
+	 * @return array
+	 */
+	public static function flatten_inner_blocks( $inner_blocks, $filter_options, $flattened_blocks = [] ) {
+		foreach ( $inner_blocks as $inner_block ) {
+			// This block doesnt have any inner blocks, so just add it to the flattened blocks. Ensure the parentId is set.
+			if ( ! isset( $inner_block['innerBlocks'] ) ) {
+				$inner_block['parentId'] = $inner_block['parentId'] ?? $filter_options['parentId'];
+				array_push( $flattened_blocks, $inner_block );
+				// This block is a root block, so go through the inner blocks recursively.
+			} elseif ( ! isset( $inner_block['parentId'] ) ) {
+				$inner_blocks_copy = $inner_block['innerBlocks'];
+				unset( $inner_block['innerBlocks'] );
+				$inner_block['parentId'] = $inner_block['parentId'] ?? $filter_options['parentId'];
+				array_push( $flattened_blocks, $inner_block );
+				$flattened_blocks = self::flatten_inner_blocks( $inner_blocks_copy, $filter_options, $flattened_blocks );
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockDataAttribute',
+			[
+				'description' => 'Block data attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => 'String',
+						'description' => 'Block data attribute name',
+					],
+					'value' => [
+						'type'        => 'String',
+						'description' => 'Block data attribute value',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual inner block, with the above attribute.
+		register_graphql_type(
+			'InnerBlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'         => [
+						'type'        => 'String',
+						'description' => 'ID of the block',
+					],
+					'parentId'   => [
+						'type'        => 'String',
+						'description' => 'ID of the parent for this inner block, if it is an inner block. This will match the ID of the block',
+					],
+					'name'       => [
+						'type'        => 'String',
+						'description' => 'Block name',
+					],
+					'attributes' => [
+						'type'        => [
+							'list_of' => 'BlockDataAttribute',
+						],
+						'description' => 'Block data attributes',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual block, with the above attribute.
+		register_graphql_type(
+			'BlockData',
```

Comment:
It seems like you don't need two distinct types for Blocks and InnerBlocks. You could have the same type with nullable fields for `parentId` and `innerBlocks`?

Reply:
The reason I had kept it separate was because the outer blocks will have only `innerBlocks` but the innerBlocks will only `parentId`.  I didn't want to potentially allow a query where you could build `innerBlocks` at any depth. This way you are restricted to what we want, while still having some flexibility like you would in GraphQL land.

Were you suggesting that by doing it the way you mentioned, this would not be a concern? Hope I understood that right.

Reply:
I basically just saying use the same type for both inner and outer blocks and use `null` appropriately, like:

```json
{
   "blockData": [
      {
         "id": "abc123=",
         "parentId": null,
         "name": "core/some-outer-block",
         "attributes": [],
         "innerBlocks": [
            {
               "id": "def456=",
               "parentId": "abc123=",
               "name": "core/some-outer-block",
               "attributes": [],
               "innerBlocks": null
            }
         ]
      }
   ]
}
```

Reply:
Yip, I understand what you mean. @alecgeatches had suggested something similar and I had resisted doing it but you are both right on this. It would make it simpler overall, so I'll make that change.

Reply:
I have implemented this change now, so it's a shared type with the nullable set for parentId and innerBlocks. In case I haven't done the nullable part right, lemme know.